man/man8/swtpm_setup.8
.\" Automatically generated by Pod::Man 4.12 (Pod::Simple 3.39)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "swtpm_setup 8"
.TH swtpm_setup 8 "2020-08-20" "swtpm" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
swtpm_setup \- Swtpm utility to simulate the manufacturing of a TPM 1.2 or 2.0
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBswtpm_setup [\s-1OPTIONS\s0]\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBswtpm_setup\fR is a tool that prepares the initial state for a libtpms-based
\&\s-1TPM.\s0
.PP
For creating the initial state of a \s-1TPM 1.2,\s0 swtpm_setup must be run either
as root or as the user the that tcsd requires, which is typically tss.
.PP
The following options are supported:
.IP "\fB\-\-runas <userid\fR>" 4
.IX Item "--runas <userid>"
Use this userid to run swtpm_setup.sh; by default 'tss' is used.
.IP "\fB\-\-config <file\fR>" 4
.IX Item "--config <file>"
Path to configuration file containing the tool to use for creating
certificates; see also \fBswtpm_setup.conf\fR
.Sp
If this parameter is not provided, the default configuration file
/etc/swtpm_setup.conf will be used. If the environment variable
\&\s-1XDG_CONFIG_HOME\s0 is set, the configuration file is assumed to be
\&\f(CW$XDG_CONFIG_HOME\fR/swtpm_setup.conf.
.IP "\fB\-\-tpm\-state <dir\fR> or \fB\-\-tpmstate <dir\fR>" 4
.IX Item "--tpm-state <dir> or --tpmstate <dir>"
Path to a directory where the \s-1TPM\s0's state will be written into;
this is a mandatory argument
.IP "\fB\-\-tpm <path to executable\fR>" 4
.IX Item "--tpm <path to executable>"
Path to the \s-1TPM\s0 executable; this is an optional argument and
by default the swtpm executable found in the \s-1PATH\s0 will be used.
.IP "\fB\-\-tpm2\fR" 4
.IX Item "--tpm2"
Do setup on a \s-1TPM 2\s0; by default a \s-1TPM 1.2\s0 is setup.
.IP "\fB\-\-createek\fR" 4
.IX Item "--createek"
Create the \s-1EK\s0
.IP "\fB\-\-allow\-signing\fR" 4
.IX Item "--allow-signing"
Create an \s-1EK\s0 that can sign. This option requires \-\-tpm2.
.Sp
Note that the \s-1TCG\s0 specification \*(L"\s-1EK\s0 Credential Profile For \s-1TPM\s0 Family 2.0; Level 0\*(R"
suggests in its section on \*(L"\s-1EK\s0 Usage\*(R" that \*(L"the Endorsement Key can be a
created as a decryption or signing key.\*(R" However, some platforms will
not accept an \s-1EK\s0 as a signing key, or as a signing and encryption key, and
therefore this option should be used very carfully.
.IP "\fB\-\-decryption\fR" 4
.IX Item "--decryption"
Create an \s-1EK\s0 that can be used for key encipherment. This is the default
unless \-\-allow\-signing is passed. This option requires \-\-tpm2.
.IP "\fB\-\-ecc\fR" 4
.IX Item "--ecc"
Create elliptic curve crypto (\s-1ECC\s0) keys; by default \s-1RSA\s0 keys are generated.
.IP "\fB\-\-take\-ownership\fR" 4
.IX Item "--take-ownership"
Take ownership; this option implies \-\-createek
.IP "\fB\-\-ownerpass  <password\fR>" 4
.IX Item "--ownerpass <password>"
Provide custom owner password; default is ooo
.IP "\fB\-\-owner\-well\-known\fR" 4
.IX Item "--owner-well-known"
Use a password of all zeros (20 bytes of zeros) as the owner password
.IP "\fB\-\-srkpass <password\fR>" 4
.IX Item "--srkpass <password>"
Provide custom \s-1SRK\s0 password; default is sss
.IP "\fB\-\-srk\-well\-known\fR" 4
.IX Item "--srk-well-known"
Use a password of all zeros (20 bytes of zeros) as the \s-1SRK\s0 password
.IP "\fB\-\-create\-ek\-cert\fR" 4
.IX Item "--create-ek-cert"
Create an \s-1EK\s0 certificate; this implies \-\-createek
(\s-1NOT SUPPORTED YET\s0)
.IP "\fB\-\-create\-platform\-cert\fR" 4
.IX Item "--create-platform-cert"
Create a platform certificate; this implies \-\-create\-ek\-cert
.IP "\fB\-\-lock\-nvram\fR" 4
.IX Item "--lock-nvram"
Lock \s-1NVRAM\s0 access
.IP "\fB\-\-display\fR" 4
.IX Item "--display"
At the end display as much info as possible about the configuration
of the \s-1TPM\s0
.IP "\fB\-\-logfile <logfile\fR>" 4
.IX Item "--logfile <logfile>"
The logfile to log to. By default logging goes to stdout and stderr.
.IP "\fB\-\-keyfile <keyfile\fR>" 4
.IX Item "--keyfile <keyfile>"
The key file contains an \s-1ASCII\s0 hex key consisting of 32 hex digits with an
optional leading '0x'. This is the key to be used by the \s-1TPM\s0 emulator
for encrypting the state of the \s-1TPM.\s0
.IP "\fB\-\-keyfile\-fd <file descriptor\fR>" 4
.IX Item "--keyfile-fd <file descriptor>"
Like \fB\-\-keyfile\fR but the key will be read from the file descriptor.
.IP "\fB\-\-pwdfile <passphrase file\fR>" 4
.IX Item "--pwdfile <passphrase file>"
The passphrase file contains a passphrase from which the \s-1TPM\s0 emulator
will derive the encyrption key from and use the key for encrypting the \s-1TPM\s0
state.
.IP "\fB\-\-pwdfile\-fd <file descriptor\fR>" 4
.IX Item "--pwdfile-fd <file descriptor>"
Like \fB\-\-pwdfile\fR but the passphrase will be read from the file descriptor.
.IP "\fB\-\-ciper <cipher\fR>" 4
.IX Item "--ciper <cipher>"
The cipher may be either aes-cbc or aes\-128\-cbc for 128 bit \s-1AES\s0 encryption,
or aes\-256\-cbc for 256 bit \s-1AES\s0 encryption. The same cipher must be used
on the \fIswtpm\fR command line later on.
.IP "\fB\-\-overwrite\fR" 4
.IX Item "--overwrite"
Overwrite existing \s-1TPM\s0 state. All previous state will be erased.
If this option is not given and an existing state file is found, an error
code is returned.
.IP "\fB\-\-not\-overwrite\fR" 4
.IX Item "--not-overwrite"
Do not overwrite existing \s-1TPM\s0 state. If existing \s-1TPM\s0 state is found, the
program ends without an error.
.IP "\fB\-\-vmid <\s-1VM ID\s0\fR>" 4
.IX Item "--vmid <VM ID>"
Optional \s-1VM ID\s0 that can be used to keep track of certificates issued
for VMs (or containers). This parameter will be passed through to the tool
used for creating the certificates and may be required by that tool.
.IP "\fB\-\-pcr\-banks <\s-1PCR\s0 banks\fR>" 4
.IX Item "--pcr-banks <PCR banks>"
Optional comma-separated list of \s-1PCR\s0 banks to activate. Providing '\-'
allows a user to skip the selection and activates all \s-1PCR\s0 banks. By default
the sha1 and sha256 banks are activated.
.IP "\fB\-\-swtpm_ioctl <executable\fR>" 4
.IX Item "--swtpm_ioctl <executable>"
Pass the path to the swtpm_ioctl executable. By default the swtpm_ioctl
in the \s-1PATH\s0 is used.
.IP "\fB\-\-tcsd\-system\-ps\-file <file\fR>" 4
.IX Item "--tcsd-system-ps-file <file>"
A file to copy \s-1TCSD\s0's system_ps_file to. The system_ps_file contains the
\&\s-1TPM 1.2 SRK\s0 public key after taking ownership of the \s-1TPM.\s0 The file is
needed by \s-1TCSD\s0 for key related functions.
.Sp
This option is only useful with \s-1TPM 1.2\s0 and in if ownership is taken.
.IP "\fB\-\-rsa\-keysize <keysize\fR> (since v0.4)" 4
.IX Item "--rsa-keysize <keysize> (since v0.4)"
This option allows to pass the size of a \s-1TPM 2 RSA EK\s0 key, such as 2048
or 3072. The supported keysizes for a \s-1TPM 2\s0 can be queried for using
the \fI\-\-print\-capabilities\fR option. The default size is 2048 bits for
both \s-1TPM 1.2\s0 and \s-1TPM 2.\s0 If 'max' is passed, the largest possible key
size is used.
.IP "\fB\-\-print\-capabilities\fR (since v0.2)" 4
.IX Item "--print-capabilities (since v0.2)"
Print capabilities that were added to swtpm_setup after version 0.1.
The output may contain the following:
.Sp
.Vb 10
\&    {
\&      "type": "swtpm_setup",
\&      "features": [
\&        "cmdarg\-keyfile\-fd",
\&        "cmdarg\-pwdfile\-fd",
\&        "no\-tpm12\-tools",
\&        "tpm2\-rsa\-keysize\-2048",
\&        "tpm2\-rsa\-keysize\-3072"
\&      ]
\&    }
.Ve
.Sp
The meaning of the feature verbs is as follows:
.RS 4
.IP "\fBcmdarg-key-fd\fR" 4
.IX Item "cmdarg-key-fd"
The \fI\-\-keyfile\-fd\fR option is supported.
.IP "\fBcmdarg-pwd-fd\fR" 4
.IX Item "cmdarg-pwd-fd"
The \fI\-\-pwdfile\-fd\fR option is supported.
.IP "\fBno\-tpm12\-tools\fR" 4
.IX Item "no-tpm12-tools"
\&\s-1TPM 1.2\s0 tools cannot be found in the \s-1PATH.\s0 This means that no certificates can
be created since they cannot be written into the \s-1NVRAM\s0 and the \s-1NVRAM\s0 cannot be
locked. Among the \s-1TPM 1.2\s0 setup parameters only \fI\-\-createek\fR can be passed.
.IP "\fBtpm2\-rsa\-keysize\-2048, ...\fR" 4
.IX Item "tpm2-rsa-keysize-2048, ..."
The shown \s-1RSA\s0 key sizes are supported for a \s-1TPM 2\s0's \s-1EK\s0 key. If none of the
tpm2\-rsa\-keysize verbs is shown then only \s-1RSA 2048\s0 bit keys are supported.
.RE
.RS 4
.RE
.IP "\fB\-\-help, \-h\fR" 4
.IX Item "--help, -h"
Display the help screen
.SH "EXAMPLE USAGE"
.IX Header "EXAMPLE USAGE"
To simulate manufacturing of a \s-1TPM,\s0 one would typically run the following command:
.PP
.Vb 2
\&  #> sudo swtpm_setup \-\-tpmstate /tmp/mytpm1/ \e
\&      \-\-create\-ek\-cert \-\-create\-platform\-cert \-\-lock\-nvram
.Ve
.PP
Note: since setting up a \s-1TPM 1.2\s0 relies on the \fItcsd\fR for some of its operations,
\&\fBswtpm_setup\fR has to be run as root so that it can invoke the \fItcsd\fR either as root
or tss user.
.PP
A normal user can also simulate the manufacturing of a \s-1TPM 2\s0 using the
swtpm-localca plugin. The following example assumes that the user has
set the environment variable \s-1XDG_CONFIG_HOME\s0 as follows (using bash for
example):
.PP
.Vb 1
\&    export XDG_CONFIG_HOME=~/.config
.Ve
.PP
Note: The \s-1XDG_CONFIG_HOME\s0 variable is part of the \s-1XDG\s0 Base Directory
Specification.
.PP
The following configuration files need to be created:
.PP
~/.config/swtpm_setup.conf:
.PP
.Vb 4
\&    # Program invoked for creating certificates
\&    create_certs_tool= /usr/share/swtpm/swtpm\-localca
\&    create_certs_tool_config = ${XDG_CONFIG_HOME}/swtpm\-localca.conf
\&    create_certs_tool_options = ${XDG_CONFIG_HOME}/swtpm\-localca.options
.Ve
.PP
~/.config/swtpm\-localca.conf:
.PP
.Vb 4
\&    statedir = ${XDG_CONFIG_HOME}/var/lib/swtpm\-localca
\&    signingkey = ${XDG_CONFIG_HOME}/var/lib/swtpm\-localca/signkey.pem
\&    issuercert = ${XDG_CONFIG_HOME}/var/lib/swtpm\-localca/issuercert.pem
\&    certserial = ${XDG_CONFIG_HOME}/var/lib/swtpm\-localca/certserial
.Ve
.PP
~/.config/swtpm\-localca.options:
.PP
.Vb 3
\&    \-\-platform\-manufacturer Fedora
\&    \-\-platform\-version 2.12
\&    \-\-platform\-model QEMU
.Ve
.PP
Note: The tool swtpm-create-user-config-files can be used to create such
files (with different content):
.PP
.Vb 4
\&  #> /usr/share/swtpm/swtpm\-create\-user\-config\-files
\&  Writing /home/stefanb/.config/swtpm_setup.conf.
\&  Writing /home/stefanb/.config/swtpm\-localca.conf.
\&  Writing /home/stefanb/.config/swtpm\-localca.options.
.Ve
.PP
The following commands now create a \s-1TPM 2\s0 with an \s-1EK\s0 and platform
certificate. The state of the \s-1TPM 2\s0 will be stored in the directory
${\s-1XDG_CONFIG_HOME\s0}/mytpm1.
.PP
.Vb 3
\&  #> mkdir \-p ${XDG_CONFIG_HOME}/mytpm1
\&  #> swtpm_setup \-\-tpm2 \-\-tpmstate ${XDG_CONFIG_HOME}/mytpm1 \e
\&      \-\-create\-ek\-cert \-\-create\-platform\-cert \-\-lock\-nvram
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBswtpm_setup.conf\fR
.SH "REPORTING BUGS"
.IX Header "REPORTING BUGS"
Report bugs to Stefan Berger <stefanb@linux.vnet.ibm.com>


CHANGES
CHANGES - changes for swtpm

version 0.3.4:
  - swtpm:
    - Fix compilation for cygwin
  - swtpm_setup & swtpm-localca:
    - Get rid of bash's eval when invoking external tools to avoid abuse.
      Only use eval for 'resolving' variables.
  - tests:
    - Various fixes of minor issues

version 0.3.3:
  - swtpm_setup:
    - openSUSE: Support tcsd configuration where tss user != tss group,
                such as root/tss; Fedora & Ubuntu for example use tss/tss
  - build-sys:
    - Check whether tss user and group are available

version 0.3.2:
  - swtpm:
    - Remove unnecessary #include <seccomp.h> (fixes SuSE build)
    - Make coverity happy by handling default case in case statement
  - swtpm_setup:
    - bugfix: Create ECC storage primary key in owner hierarchy
    - bugfix: remove tpm2_stirrandom and tpm2_changeeps
  - tests:
    - Adjusted pcrUpdateCounter in tests to succeed with PCR TCB group fixes
      in libtpms TPM 2 code

version 0.3.1:
  - swtpm:
    - Fix vtpm proxy case without startup flags

version 0.3.0:
  - swtpm:
      - Support for applying 'TPM Startup' command during initialization
      - Use writev_full rather than writev; fixes --vtpm-proxy EIO error
      - Only accept() new client ctrl connection if we have none (bugfix)
  - swtpm_setup & related scripts:
      - Support whitespaces in filenames and paths
      - Do not fail on future PCR banks' hashes
  - swtpm_cert:
      - Fix OIDs for TPM 2 platforms data
      - Option parsing cleanup
      - Support for passing password in various forms
      - Use gnutls_x509_crt_get_subject_key_id API call for subj keyId
      - Support 64bit serial numbers read from command line
  - swtpm_ioctl:
      - Block SIGPIPE so we can get EPIPE on write()
  - swtpm_bios:
      - Block SIGPIPE so we can get EPIPE on write()
  - tests:
      - Increased timeouts and better support for running tests with
        executables run by valgrind
      - Allow running tests with choice of seccomp profile option
        (SWTPM_TEST_SECCOMP_OPT) to enable building for Ubuntu
      - Various cleanups & fixes
  - SELinux:
      - More rules added for support on F30

version 0.2.0:
  - Linux: swtpm now runs with a seccomp profile (blacklist) if compiled with
           libseccomp support
  - Added subpport for passing key and passphrase via file descriptor
  - TPM 2 commands can now be prefixed by 'the TCG header' and responses will
    have a 4-byte prefix and 4-byte suffix.
  - Added --print-capabilities command line option
  - Proper handling on EINTR on read, poll, and write

version 0.1.0:
  first public release


src/swtpm_setup/swtpm_setup.c
/*
 * swtpm_setup.c
 *
 * Authors: Stefan Berger <stefanb@us.ibm.com>
 *
 * (c) Copyright IBM Corporation 2011,2014,2015.
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * Neither the names of the IBM Corporation nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdbool.h>
#include <string.h>
#include <errno.h>
#include <limits.h>
#include <libgen.h>
#include <sys/types.h>
#include <pwd.h>
#include <grp.h>
#include <fcntl.h>

#if defined __APPLE__
#include <sys/mount.h>
#include <mach-o/dyld.h>
#endif

/*
 * Those parameters interpreted by swtpm_setup.sh that have an additional
 * parameter.
 */
const char *one_arg_params[] = {
    "--tpm-state",
    "--tpmstate",
    "--tpm",
    "--ownerpass",
    "--srkpass",
    "--config",
    "--vmid",
    "--logfile",
    "--keyfile",
    "--pwdfile",
    "--swtpm_ioctl",
    "--pcr-banks",
    "--tcsd-system-ps-file",
    "--rsa-keysize",
    NULL
};

/*
 * Those parameters interpreted by swtpm_setup.sh that have a file descriptor
 * parameter.
 */
const char *fd_arg_params[] = {
    "--keyfile-fd",
    "--pwdfile-fd",
    NULL
};

/*
 * Sanitize the file descriptor we pass to swtpm_setup.sh so that it can
 * freely use any fds in the range [100..109], e.g. 'exec 100 ... '.
 */
static int move_reserved_fd(const char *fdstring, char **newfdstring)
{
    char *endptr;
    long fd;
    int newfd;

    errno = 0;
    fd = strtol(fdstring, &endptr, 10);
    if (fdstring == endptr || *endptr != '\0' || fd < 0 || errno != 0) {
        fprintf(stderr, "Invalid file descriptor '%s'.\n", fdstring);
        return -1;
    }

    /* reserve file descriptors 100 - 109 for swtpm_setup.sh to use */
    if (fd >= 100 && fd <= 109) {
        newfd = fcntl(fd, F_DUPFD, 3);
        if (newfd < 0) {
            fprintf(stderr, "F_DUPFD failed: %s\n", strerror(errno));
            return -1;
        }
        if (newfd >= 100 && newfd <= 109) {
            fprintf(stderr, "newfd is also in reserved range: %u\n", newfd);
            return -1;
        }

        close(fd);

        if (asprintf(newfdstring, "%u", newfd) < 0) {
            fprintf(stderr, "Out of memory\n");
            return -1;
        }

        return 1;
    }
    return 0;
}

static int change_process_owner(const char *user)
{
    struct passwd *passwd;
    long int uid, gid;
    char *endptr = NULL;

    uid = strtoul(user, &endptr, 10);
    if (*endptr != '\0') {
        /* a string */
        passwd = getpwnam(user);
        if (!passwd) {
            fprintf(stderr, "Error: User '%s' does not exist.\n", user);
            return -1;
        }

        if (initgroups(passwd->pw_name, passwd->pw_gid) < 0) {
            fprintf(stderr, "Error: initgroups(%s, %d) failed.\n",
                    passwd->pw_name, passwd->pw_gid);
           return -1;
        }
        gid = passwd->pw_gid;
        uid = passwd->pw_uid;
    } else {
        /* an integer */
        if ((unsigned long int)uid > UINT_MAX) {
            fprintf(stderr, "Error: uid %s outside valid range.\n", user);
            return -1;
        }
        gid = uid;
    }

    if (setgid(gid) < 0) {
        fprintf(stderr, "Error: setgid(%ld) failed.\n", gid);
        return -1;
    }
    if (setuid(uid) < 0) {
        fprintf(stderr, "Error: setuid(%ld) failed.\n", uid);
        return -1;
    }
    return 0;
}


int main(int argc, char *argv[])
{
    const char *program = "swtpm_setup.sh";
    char resolved_path[PATH_MAX];
    char *dir;
    char *path_program;
    size_t length;
    struct passwd *passwd = NULL;
    int i = 1, j;
    const char *userid = NULL;
    bool change_user = true;
    const char *p;
#if defined __APPLE__
    char path[MAXPATHLEN];
    uint32_t pathlen = sizeof(path);
#endif
    char *newargv;
    int rc;

    while (i < argc) {
        if (!strcmp("--runas", argv[i])) {
            i++;
            if (i == argc) {
                fprintf(stderr, "Missing user argument for --runas\n");
                exit(1);
            }
            userid = argv[i];
        } else if (!strcmp("--help", argv[i]) || !strcmp("-h", argv[i])) {
            change_user = false;
        } else if (!strcmp("--version", argv[i])) {
            change_user = false;
        } else if (!strcmp("--print-capabilities", argv[i])) {
            change_user = false;
        }
        for (j = 0; one_arg_params[j] != NULL; j++) {
            if (!strcmp(one_arg_params[j], argv[i])) {
                i++;
                goto skip;
            }
        }
        /* Ensure that no file descriptor overlaps with those reserved
         * for free use by swtpm_setup.sh
         */
        for (j = 0; fd_arg_params[j] != NULL; j++) {
            if (!strcmp(fd_arg_params[j], argv[i]) &&
                i + 1 < argc) {
                i++;
                rc = move_reserved_fd(argv[i], &newargv);
                switch (rc) {
                case 0:
                    /* nothing to do */
                    break;
                case 1:
                    argv[i] = newargv;
                    break;
                default:
                    return EXIT_FAILURE;
                }
                break;
            }
        }
skip:
        i++;
    }

#if defined __OpenBSD__ || defined __FreeBSD__ || defined __DragonFly__
    p = getenv("_");
#elif defined __APPLE__
    if (_NSGetExecutablePath(path, &pathlen) < 0) {
        fprintf(stderr, "Could not get path of 'self'.");
        return EXIT_FAILURE;
    }
    p = path;
#else
    p = "/proc/self/exe";
#endif
    if (!realpath(p, resolved_path)) {
        fprintf(stderr, "Could not resolve path (%s) to executable: %s\n",
                p, strerror(errno));
        return EXIT_FAILURE;
    }

    dir = dirname(resolved_path);
    if (!dir) {
        fprintf(stderr, "Could not get directory from path '%s'.",
                resolved_path);
        return EXIT_FAILURE;
    }

    length = strlen(dir) + 1 + strlen(program) + 2;

    path_program = malloc(length);
    if (!path_program) {
        fprintf(stderr, "Out of memory.\n");
        goto exit_failure;
    }

    if (snprintf(path_program, length, "%s/%s", dir, program) >=
        (int)length) {
        fprintf(stderr, "Internal error writing string.\n");
        goto exit_failure;
    }

    /*
     * Unless we saw --runas, we will not attempt to switch the user.
     */
    if (!userid)
        change_user = false;

    if (change_user && change_process_owner(userid))
        goto exit_failure;
    /*
     * need to pass unmodified argv to swtpm_setup.sh
     */
    execv(path_program, argv);

    if (passwd) {
        /* should never get here */
        fprintf(stderr, "As user %s:", passwd->pw_name);
    }

    fprintf(stderr, "Could not execute '%s' : %s\n",
            path_program, strerror(errno));

exit_failure:
    free(path_program);

    return EXIT_FAILURE;
}
